confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java

```diff
@@ -0,0 +1,379 @@
+import org.apache.iotdb.common.rpc.thrift.TRegionReplicaSet;
+import org.apache.iotdb.common.rpc.thrift.TSStatus;
+import org.apache.iotdb.confignode.client.DataNodeRequestType;
+import org.apache.iotdb.confignode.client.async.datanode.AsyncDataNodeClientPool;
+import org.apache.iotdb.confignode.client.sync.datanode.SyncDataNodeClientPool;
+import org.apache.iotdb.confignode.consensus.request.write.RemoveDataNodePlan;
+import org.apache.iotdb.confignode.consensus.request.write.UpdateRegionLocationPlan;
+import org.apache.iotdb.confignode.consensus.response.DataNodeToStatusResp;
+import org.apache.iotdb.confignode.manager.ConfigManager;
+import org.apache.iotdb.confignode.persistence.NodeInfo;
+import org.apache.iotdb.confignode.procedure.exception.ProcedureException;
+import org.apache.iotdb.confignode.procedure.scheduler.LockQueue;
+import org.apache.iotdb.mpp.rpc.thrift.TDisableDataNodeReq;
+import org.apache.iotdb.mpp.rpc.thrift.TMigrateRegionReq;
+import org.apache.iotdb.rpc.TSStatusCode;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class DataNodeRemoveHandler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataNodeRemoveHandler.class);
+
+  private ConfigManager configManager;
+
+  /** region migrate lock */
+  private final LockQueue regionMigrateLock = new LockQueue();
+
+  public DataNodeRemoveHandler(ConfigManager configManager) {
+    this.configManager = configManager;
+  }
+
+  /**
+   * Get all consensus group id in this node
+   *
+   * @param dataNodeLocation data node location
+   * @return group id list
+   */
+  public List<TConsensusGroupId> getDataNodeRegionIds(TDataNodeLocation dataNodeLocation) {
+    return configManager.getPartitionManager().getAllReplicaSets().stream()
+        .filter(rg -> rg.getDataNodeLocations().contains(dataNodeLocation))
+        .filter(rg -> rg.regionId.getType() != TConsensusGroupType.PartitionRegion)
+        .map(TRegionReplicaSet::getRegionId)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * broadcast these datanode in RemoveDataNodeReq are disabled, so they will not accept read/write
+   * request
+   *
+   * @param disabledDataNode TDataNodeLocation
+   */
+  public TSStatus broadcastDisableDataNode(TDataNodeLocation disabledDataNode) {
+    LOGGER.info(
+        "DataNodeRemoveService start send disable the Data Node to cluster, {}", disabledDataNode);
+    TSStatus status = new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
+    List<TEndPoint> otherOnlineDataNodes =
+        configManager.getLoadManager().getOnlineDataNodes(-1).stream()
+            .map(TDataNodeConfiguration::getLocation)
+            .filter(loc -> !loc.equals(disabledDataNode))
+            .map(TDataNodeLocation::getInternalEndPoint)
+            .collect(Collectors.toList());
```

Comment:
It seems that the annotation is not match with the inner code logic. By the way, why should the other onlineDataNodes need to know that there is a DataNode is removed?

Reply:
Other datanodes do not send requests to the datanodes to be deleted in a timely manner.

Reply:
This logic is the same as the previous logic. I just added it to Procedure, and then changed it later if necessary.